src/mlpack/methods/neat/acyclic_net.hpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * @file acyclic_net.hpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * Definition of the AcyclicNet class, which represents an acyclic neural
+ * network.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_NEAT_ACYCLIC_NET_HPP
+#define MLPACK_METHODS_NEAT_ACYCLIC_NET_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack{
+namespace neat /** NeuroEvolution of Augmenting Topologies */{
+/**
+ * A class representation of an acyclic neural network. Genomes are decoded
+ * into this class if the `isAcyclic` parameter is set to `true`. It is used
+ * to evaluate an input to the genome.
+ * 
+ * The steps of the decoding and activating are as follows:
+ * 1. The nodes are each assigned a "depth" depending on the number of "jumps"
+ *    it takes to reach it from an input node. If there are two paths to the
+ *    node, the longer path is used to find the depth.
+ * 2. Based on the depth, the nodes are divided into layers.
+ * 3. Starting from layer zero, the layers are activated one by one.
+ * 4. The output is taken from the last layer, i.e. the output nodes.
+ * 
+ * @tparam ActivationFunction The activation function. 
+ */
+template <class ActivationFunction>
+class AcyclicNet
+{
+ public:
+  /**
+   * Creates an AcyclicNet object.
+   * 
+   * @param DirectedGraph A map of maps storing connection genes, whose first
+   *    key is the source ID, and second key is the target ID.
+   * @param nodeDepths A vector storing the depths of each node.
+   * @param actFn The activation function.
+   * @param nodeCount The number of nodes.
+   * @param inputNodeCount The number of input nodes.
+   * @param outputNodeCount The number of output nodes.
+   * @param bias The bias.
+   */
+  AcyclicNet(std::map<size_t, std::map<size_t, ConnectionGene>>& directedGraph,
+             std::vector<size_t> nodeDepths,
+             ActivationFunction& actFn,
+             const size_t nodeCount,
+             const size_t inputNodeCount,
+             const size_t outputNodeCount,
+             const double bias);
+
+  /**
+   * Finds the output of the network from the given input.
+   */
+  arma::vec Evaluate(arma::vec input);
+
+ private:
+  /**
+   * A digraph containing connection genes sorted by source ID, and then
+   * secondary sorted by target ID.
+   */
+  std::map<size_t, std::map<size_t, ConnectionGene>> directedGraph;
```

Comment:
I am wondering, do we really need to store directedGraph here, I mean we can always pass the graph with evaluate right? Let me know your thoughts.

Reply:
Good idea, I missed that. I'll see where else I can do something like this as well.